File: Tests/RedlineCoreTests/SnapshotTests.swift
import XCTest
@testable import RedlineCore

final class SnapshotTests: XCTestCase {
    private var dir: URL!
    private var file: URL!

    override func setUpWithError() throws {
        dir = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("redline-snap-" + UUID().uuidString)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        file = dir.appendingPathComponent("snapshot.json")
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: dir)
    }

    private func sample(updatedAt: Date = Date()) -> Snapshot {
        var agg = Agg()
        agg.io = 6044
        agg.cost = 1.25
        agg.hasUnpriced = true
        let limits = [
            LimitWindow(provider: "Claude", key: "five_hour", utilization: 12,
                        resetsAt: Date(timeIntervalSince1970: 4_000_000_000)),
            LimitWindow(provider: "Codex", key: "five_hour", utilization: 41, resetsAt: nil),
            LimitWindow(provider: "Claude", key: "seven_day", utilization: 8, resetsAt: nil),
        ]
        return Snapshot(updatedAt: updatedAt, limits: limits, today: agg, week: agg)
    }

    func testRoundTripsThroughDisk() {
        let snap = sample()
        XCTAssertTrue(SnapshotStore.write(snap, to: file))
        let back = SnapshotStore.read(from: file)
        XCTAssertEqual(back, snap, "dates must survive the ISO8601 round trip")
    }

    /// Claude's windows age on their own clock: the feed writes only while Claude Code runs,
    /// so their stamp must survive the trip and drive staleness independently of updatedAt.
    func testClaudeLimitsAsOfRoundTripsAndAges() throws {
        let now = Date(timeIntervalSince1970: 1_755_400_000)
        var agg = Agg()
        agg.io = 1
        let snap = Snapshot(updatedAt: now,
                            limits: [LimitWindow(provider: "Claude", key: "seven_day",
                                                 utilization: 30, resetsAt: nil)],
                            today: agg, week: agg,
                            claudeLimitsAsOf: now.addingTimeInterval(-1200))
        XCTAssertTrue(SnapshotStore.write(snap, to: file))
        let back = try XCTUnwrap(SnapshotStore.read(from: file))
        XCTAssertEqual(back.claudeLimitsAsOf, snap.claudeLimitsAsOf)
        XCTAssertFalse(back.isStale(now: now), "the snapshot itself is current")
        XCTAssertTrue(back.claudeLimitsAreStale(now: now),
                      "while its Claude windows are twenty minutes old")
        XCTAssertFalse(back.claudeLimitsAreStale(now: now, tolerance: 1800))
    }

    /// Snapshots written before the field existed decode with no stamp, and no stamp must
    /// read as "nothing to age" rather than stale.
    func testOlderSnapshotsWithoutTheStampStillDecode() throws {
        let snap = sample()
        XCTAssertTrue(SnapshotStore.write(snap, to: file))
        var json = try JSONSerialization.jsonObject(
            with: Data(contentsOf: file)) as! [String: Any]
        json.removeValue(forKey: "claudeLimitsAsOf")
        try JSONSerialization.data(withJSONObject: json).write(to: file)
        let back = try XCTUnwrap(SnapshotStore.read(from: file))
        XCTAssertNil(back.claudeLimitsAsOf)
        XCTAssertFalse(back.claudeLimitsAreStale())
    }

    func testTimestampsAreTruncatedToWholeSeconds() {
        // ISO8601 cannot carry sub-second precision, so the initializer normalizes rather
        // than leaving an instance that fails to equal its own round trip.
        let odd = Date(timeIntervalSince1970: 1_700_000_000.987)
        XCTAssertEqual(sample(updatedAt: odd).updatedAt.timeIntervalSince1970,
                       1_700_000_000)
    }

    func testWorstPicksHighestUtilizationAcrossProviders() {
        XCTAssertEqual(sample().worst(prefix: "five_hour")?.provider, "Codex")
        XCTAssertEqual(sample().worst(prefix: "five_hour")?.utilization, 41)
        XCTAssertNil(sample().worst(prefix: "nope"))
    }

    func testCarriesUnpricedFlagSoTheWidgetCanMarkPartialCost() {
        XCTAssertTrue(sample().today.hasUnpriced)
    }

    func testStalenessIsDetectable() {
        let now = Date()
        let fresh = sample(updatedAt: now.addingTimeInterval(-60))
        let old = sample(updatedAt: now.addingTimeInterval(-3600))
        XCTAssertFalse(fresh.isStale(now: now))
        XCTAssertTrue(old.isStale(now: now),
                      "a widget must be able to say the reading is not current")
    }

    func testReadingAbsentFileReturnsNil() {
        XCTAssertNil(SnapshotStore.read(from: dir.appendingPathComponent("absent.json")))
    }

    func testFallsBackToUserPathWithoutAnAppGroup() {
        let url = SnapshotStore.url(appGroup: nil)
        XCTAssertTrue(url.path.hasSuffix(".local/share/redline/snapshot.json"))
    }
}

final class SnapshotLocationTests: XCTestCase {
    func testWidgetContainerPathIsInsideTheWidgetsOwnSandbox() {
        let p = SnapshotStore.widgetContainerURL.path
        XCTAssertTrue(p.contains("Library/Containers/\(SnapshotStore.widgetBundleID)/Data"),
                      "must address the widget's container, the one place it can always read")
        XCTAssertTrue(p.hasSuffix("redline/snapshot.json"))
    }

    func testOwnContainerIsTriedFirstWhenReading() {
        guard let first = SnapshotStore.readCandidates.first else {
            return XCTFail("expected at least one read candidate")
        }
        XCTAssertEqual(first, SnapshotStore.localAppSupportURL,
                       "a sandboxed widget can only rely on its own container")
    }

    func testUserPathIsAlwaysAWriteTarget() {
        XCTAssertTrue(SnapshotStore.writeTargets.contains(SnapshotStore.userURL))
    }
}

final class SnapshotProviderViewTests: XCTestCase {
    private func make() -> Snapshot {
        var today = Agg()
        today.io = 1000
        today.providers = ["Claude": ProviderUsage(io: 900, cost: 9, models: [:]),
                           "Codex": ProviderUsage(io: 100, cost: 0, models: [:])]
        var week = Agg()
        week.io = 5000
        week.providers = ["Claude": ProviderUsage(io: 5000, cost: 50, models: [:])]
        let limits = [
            LimitWindow(provider: "Claude", key: "five_hour", utilization: 10, resetsAt: nil),
            LimitWindow(provider: "Claude", key: "seven_day", utilization: 20, resetsAt: nil),
            LimitWindow(provider: "Codex", key: "seven_day", utilization: 80, resetsAt: nil),
        ]
        return Snapshot(updatedAt: Date(), limits: limits, today: today, week: week,
                        ollama: Snapshot.Ollama(
                            reachable: true, version: "0.5.0",
                            running: [.init(name: "qwen3:30b", sizeBytes: 100,
                                            vramShare: 0.5)],
                            downloadedCount: 3, downloadedBytes: 30_000_000_000))
    }

    func testWorstScopedToOneProvider() {
        let s = make()
        XCTAssertEqual(s.worst(prefix: "seven_day")?.utilization, 80, "unscoped takes the max")
        XCTAssertEqual(s.worst(prefix: "seven_day", provider: "Claude")?.utilization, 20,
                       "scoped must ignore other providers")
    }

    func testWindowsForProvider() {
        XCTAssertEqual(make().windows(for: "Codex").count, 1)
        XCTAssertTrue(make().windows(for: "Ollama").isEmpty)
    }

    func testPerProviderTotals() {
        let s = make()
        XCTAssertEqual(s.today(for: "Claude").io, 900)
        XCTAssertEqual(s.today(for: nil).io, 1000, "nil means every provider")
        XCTAssertEqual(s.week(for: "Codex").io, 0, "absent provider reports zero, not a crash")
    }

    func testOllamaSectionSurvivesTheRoundTrip() throws {
        let dir = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("redline-\(UUID().uuidString).json")
        defer { try? FileManager.default.removeItem(at: dir) }
        XCTAssertTrue(SnapshotStore.write(make(), to: dir))
        let back = SnapshotStore.read(from: dir)
        XCTAssertEqual(back?.ollama?.version, "0.5.0")
        XCTAssertEqual(back?.ollama?.running.first?.vramShare, 0.5)
        XCTAssertEqual(back?.ollama?.downloadedCount, 3)
    }

    func testOlderSnapshotWithoutNewFieldsStillDecodes() throws {
        let dir = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("redline-\(UUID().uuidString).json")
        defer { try? FileManager.default.removeItem(at: dir) }
        // A file written before per-provider totals and Ollama existed
        let legacy = """
        {"updatedAt":"2026-08-12T10:00:00Z","limits":[],
         "today":{"io":1,"cost":0,"hasUnpriced":false},
         "week":{"io":2,"cost":0,"hasUnpriced":false}}
        """
        try Data(legacy.utf8).write(to: dir)
        let back = SnapshotStore.read(from: dir)
        XCTAssertNotNil(back, "a widget must not go blank because the format grew")
        XCTAssertNil(back?.ollama)
        XCTAssertEqual(back?.today(for: "Claude").io, 0)
    }
}

final class SnapshotWindowIdentityTests: XCTestCase {
    func testWindowIdIncludesProvider() {
        let claude = Snapshot.Window(provider: "Claude", key: "seven_day",
                                     utilization: 5, resetsAt: nil)
        let codex = Snapshot.Window(provider: "Codex", key: "seven_day",
                                    utilization: 17, resetsAt: nil)
        XCTAssertNotEqual(claude.id, codex.id,
                          "a shared key rendered one provider twice and dropped the other")
    }

    func testWindowIdsAreUniqueAcrossATypicalSnapshot() {
        let windows = [
            Snapshot.Window(provider: "Claude", key: "five_hour", utilization: 8, resetsAt: nil),
            Snapshot.Window(provider: "Claude", key: "seven_day", utilization: 5, resetsAt: nil),
            Snapshot.Window(provider: "Codex", key: "seven_day", utilization: 17, resetsAt: nil),
        ]
        XCTAssertEqual(Set(windows.map(\.id)).count, windows.count)
    }

    func testWindowDisplayNameIsReadableAndProviderAware() {
        XCTAssertEqual(Snapshot.Window(provider: "Claude", key: "five_hour",
                                       utilization: 0, resetsAt: nil).displayName,
                       "Session · 5h")
        XCTAssertEqual(Snapshot.Window(provider: "Codex", key: "seven_day",
                                       utilization: 0, resetsAt: nil).displayName, "Week")
        XCTAssertEqual(Snapshot.Window(provider: "Claude", key: "seven_day",
                                       utilization: 0, resetsAt: nil).displayName,
                       "Week · all models")
    }
}

File: windows/tests/Redline.Core.Tests/SnapshotTests.cs
using System.Text.Json.Nodes;
using Redline.Core;

namespace Redline.Core.Tests;

public sealed class SnapshotTests : IDisposable
{
    private readonly string _dir;
    private readonly string _file;

    public SnapshotTests()
    {
        _dir = Path.Combine(Path.GetTempPath(), "redline-snap-" + Guid.NewGuid());
        Directory.CreateDirectory(_dir);
        _file = Path.Combine(_dir, "snapshot.json");
    }

    public void Dispose()
    {
        try { Directory.Delete(_dir, true); } catch { }
    }

    private static Snapshot Sample(DateTimeOffset? updatedAt = null)
    {
        var agg = new Agg { Io = 6044, Cost = 1.25, HasUnpriced = true };
        var limits = new[]
        {
            new LimitWindow("Claude", "five_hour", 12, DateTimeOffset.FromUnixTimeSeconds(4_000_000_000)),
            new LimitWindow("Codex", "five_hour", 41, null),
            new LimitWindow("Claude", "seven_day", 8, null),
        };
        return new Snapshot(updatedAt ?? DateTimeOffset.UtcNow, limits, agg, agg);
    }

    [Fact]
    public void RoundTripsThroughDisk()
    {
        var snap = Sample();
        Assert.True(SnapshotStore.Write(snap, _file));
        var back = SnapshotStore.Read(_file);
        // Dates must survive the ISO8601 round trip
        Assert.Equal(snap, back);
    }

    /// Claude's windows age on their own clock: the feed writes only while Claude Code runs,
    /// so their stamp must survive the trip and drive staleness independently of updatedAt.
    [Fact]
    public void ClaudeLimitsAsOfRoundTripsAndAges()
    {
        var now = DateTimeOffset.FromUnixTimeSeconds(1_755_400_000);
        var agg = new Agg { Io = 1 };
        var snap = new Snapshot(now, new[] { new LimitWindow("Claude", "seven_day", 30, null) },
                                agg, agg, claudeLimitsAsOf: now.AddSeconds(-1200));
        Assert.True(SnapshotStore.Write(snap, _file));
        var back = SnapshotStore.Read(_file);
        Assert.NotNull(back);
        Assert.Equal(snap.ClaudeLimitsAsOf, back!.ClaudeLimitsAsOf);
        // The snapshot itself is current
        Assert.False(back.IsStale(now));
        // While its Claude windows are twenty minutes old
        Assert.True(back.ClaudeLimitsAreStale(now));
        Assert.False(back.ClaudeLimitsAreStale(now, tolerance: 1800));
    }

    /// Snapshots written before the field existed decode with no stamp, and no stamp must
    /// read as "nothing to age" rather than stale.
    [Fact]
    public void OlderSnapshotsWithoutTheStampStillDecode()
    {
        var snap = Sample();
        Assert.True(SnapshotStore.Write(snap, _file));
        var json = JsonNode.Parse(File.ReadAllText(_file))!.AsObject();
        json.Remove("claudeLimitsAsOf");
        File.WriteAllText(_file, json.ToJsonString());
        var back = SnapshotStore.Read(_file);
        Assert.NotNull(back);
        Assert.Null(back!.ClaudeLimitsAsOf);
        Assert.False(back.ClaudeLimitsAreStale());
    }

    [Fact]
    public void TimestampsAreTruncatedToWholeSeconds()
    {
        // ISO8601 carries no sub-second precision, so the constructor normalizes
        var odd = DateTimeOffset.FromUnixTimeMilliseconds(1_700_000_000_987);
        Assert.Equal(1_700_000_000, Sample(odd).UpdatedAt.ToUnixTimeSeconds());
        Assert.Equal(0, Sample(odd).UpdatedAt.Millisecond);
    }

    [Fact]
    public void WorstPicksHighestUtilizationAcrossProviders()
    {
        Assert.Equal("Codex", Sample().Worst("five_hour")?.Provider);
        Assert.Equal(41, Sample().Worst("five_hour")?.Utilization);
        Assert.Null(Sample().Worst("nope"));
    }

    [Fact]
    public void CarriesUnpricedFlagSoTheWidgetCanMarkPartialCost()
    {
        Assert.True(Sample().Today.HasUnpriced);
    }

    [Fact]
    public void StalenessIsDetectable()
    {
        var now = DateTimeOffset.UtcNow;
        var fresh = Sample(now.AddSeconds(-60));
        var old = Sample(now.AddSeconds(-3600));
        Assert.False(fresh.IsStale(now));
        // A widget must be able to say the reading is not current
        Assert.True(old.IsStale(now));
    }

    [Fact]
    public void ReadingAbsentFileReturnsNil()
    {
        Assert.Null(SnapshotStore.Read(Path.Combine(_dir, "absent.json")));
    }

    [Fact]
    public void FallsBackToUserPathWithoutAnAppGroup()
    {
        var path = SnapshotStore.Url(appGroup: null).Replace('\\', '/');
        Assert.EndsWith(".local/share/redline/snapshot.json", path);
    }

    /// Windows addition: the keys a Swift-written file uses, so either side reads the other's.
    [Fact]
    public void KeysMatchSwiftCodable()
    {
        var json = Sample(DateTimeOffset.FromUnixTimeSeconds(1_700_000_000)).ToJson();
        Assert.Equal("2023-11-14T22:13:20Z", Json.Str(json["updatedAt"]));
        Assert.Equal(new[] { "updatedAt", "limits", "today", "week", "todayByProvider", "weekByProvider" },
                     json.Select(kv => kv.Key));
        var first = json["limits"]![0]!.AsObject();
        Assert.Equal(new[] { "provider", "key", "utilization", "resetsAt" }, first.Select(kv => kv.Key));
        Assert.Equal(new[] { "io", "cost", "hasUnpriced" }, json["today"]!.AsObject().Select(kv => kv.Key));
        // Nil optionals are omitted, as JSONEncoder does
        Assert.Null(json["limits"]![1]!["resetsAt"]);
    }

    /// Swift's ISO8601 decoder rejects fractional seconds and missing required keys alike.
    [Fact]
    public void RequiredKeysAreRequired()
    {
        File.WriteAllText(_file, """{"updatedAt":"2026-08-12T10:00:00Z","limits":[]}""");
        Assert.Null(SnapshotStore.Read(_file));
    }
}

public sealed class SnapshotLocationTests
{
    [Fact]
    public void WidgetContainerPathIsInsideTheWidgetsOwnSandbox()
    {
        var p = SnapshotStore.WidgetContainerUrl.Replace('\\', '/');
        // Must address the widget's container, the one place it can always read
        Assert.Contains($"Library/Containers/{SnapshotStore.WidgetBundleID}/Data", p);
        Assert.EndsWith("redline/snapshot.json", p);
    }

    [Fact]
    public void OwnContainerIsTriedFirstWhenReading()
    {
        var first = SnapshotStore.ReadCandidates.FirstOrDefault();
        Assert.NotNull(first);
        // A sandboxed widget can only rely on its own container
        Assert.Equal(SnapshotStore.LocalAppSupportUrl, first);
    }

    [Fact]
    public void UserPathIsAlwaysAWriteTarget()
    {
        Assert.Contains(SnapshotStore.UserUrl, SnapshotStore.WriteTargets);
    }
}

public sealed class SnapshotProviderViewTests
{
    private static Snapshot Make()
    {
        var today = new Agg { Io = 1000 };
        today.Providers["Claude"] = new ProviderUsage { Io = 900, Cost = 9 };
        today.Providers["Codex"] = new ProviderUsage { Io = 100, Cost = 0 };
        var week = new Agg { Io = 5000 };
        week.Providers["Claude"] = new ProviderUsage { Io = 5000, Cost = 50 };
        var limits = new[]
        {
            new LimitWindow("Claude", "five_hour", 10, null),
            new LimitWindow("Claude", "seven_day", 20, null),
            new LimitWindow("Codex", "seven_day", 80, null),
        };
        return new Snapshot(DateTimeOffset.UtcNow, limits, today, week,
            ollama: new Snapshot.OllamaSection(true, "0.5.0",
                new[] { new Snapshot.OllamaSection.RunningModel("qwen3:30b", 100, 0.5) },
                3, 30_000_000_000));
    }

    [Fact]
    public void WorstScopedToOneProvider()
    {
        var s = Make();
        // Unscoped takes the max
        Assert.Equal(80, s.Worst("seven_day")?.Utilization);
        // Scoped must ignore other providers
        Assert.Equal(20, s.Worst("seven_day", provider: "Claude")?.Utilization);
    }

    [Fact]
    public void WindowsForProvider()
    {
        Assert.Single(Make().Windows("Codex"));
        Assert.Empty(Make().Windows("Ollama"));
    }

    [Fact]
    public void PerProviderTotals()
    {
        var s = Make();
        Assert.Equal(900, s.TodayFor("Claude").Io);
        // Null means every provider
        Assert.Equal(1000, s.TodayFor(null).Io);
        // An absent provider reports zero, not a crash
        Assert.Equal(0, s.WeekFor("Codex").Io);
    }

    [Fact]
    public void OllamaSectionSurvivesTheRoundTrip()
    {
        var path = Path.Combine(Path.GetTempPath(), $"redline-{Guid.NewGuid()}.json");
        try
        {
            Assert.True(SnapshotStore.Write(Make(), path));
            var back = SnapshotStore.Read(path);
            Assert.Equal("0.5.0", back?.Ollama?.Version);
            Assert.Equal(0.5, back?.Ollama?.Running.First().VramShare);
            Assert.Equal(3, back?.Ollama?.DownloadedCount);
        }
        finally { File.Delete(path); }
    }

    [Fact]
    public void OlderSnapshotWithoutNewFieldsStillDecodes()
    {
        var path = Path.Combine(Path.GetTempPath(), $"redline-{Guid.NewGuid()}.json");
        try
        {
            // A file written before per-provider totals and Ollama existed
            File.WriteAllText(path, """
                {"updatedAt":"2026-08-12T10:00:00Z","limits":[],
                 "today":{"io":1,"cost":0,"hasUnpriced":false},
                 "week":{"io":2,"cost":0,"hasUnpriced":false}}
                """);
            var back = SnapshotStore.Read(path);
            // A widget must not go blank because the format grew
            Assert.NotNull(back);
            Assert.Null(back!.Ollama);
            Assert.Equal(0, back.TodayFor("Claude").Io);
        }
        finally { File.Delete(path); }
    }
}

public sealed class SnapshotWindowIdentityTests
{
    [Fact]
    public void WindowIdIncludesProvider()
    {
        var claude = new Snapshot.Window("Claude", "seven_day", 5, null);
        var codex = new Snapshot.Window("Codex", "seven_day", 17, null);
        // A shared key rendered one provider twice and dropped the other
        Assert.NotEqual(claude.Id, codex.Id);
    }

    [Fact]
    public void WindowIdsAreUniqueAcrossATypicalSnapshot()
    {
        var windows = new[]
        {
            new Snapshot.Window("Claude", "five_hour", 8, null),
            new Snapshot.Window("Claude", "seven_day", 5, null),
            new Snapshot.Window("Codex", "seven_day", 17, null),
        };
        Assert.Equal(windows.Length, windows.Select(w => w.Id).Distinct().Count());
    }

    [Fact]
    public void WindowDisplayNameIsReadableAndProviderAware()
    {
        Assert.Equal("Session · 5h", new Snapshot.Window("Claude", "five_hour", 0, null).DisplayName);
        Assert.Equal("Week", new Snapshot.Window("Codex", "seven_day", 0, null).DisplayName);
        Assert.Equal("Week · all models", new Snapshot.Window("Claude", "seven_day", 0, null).DisplayName);
    }
}
